design/one-pager-change-logs.md

```diff
@@ -0,0 +1,310 @@
+visibility into these types of "unexpected" changes that Crossplane is making on
+their behalf.
+
+With Crossplane acting continuously and autonomously to update critical
+infrastructure, it is vital for users to have insight into the operations being
+performed, so they can build and maintain a strong sense of confidence and trust
+in their control planes.
+
+There are many approaches that can (and should) be taken to build trust along
+the various phases of the lifecycle of control planes and the infrastructure
+that they manage, but this design will focus specifically on the observability
+of changes that the control plane makes to external systems at runtime. This is
+not intended to facilitate review or approve the changes before they are made,
+but rather to provide an auditable record of all changes that have been
+performed.
+
+This means that Crossplane (and its Providers) should start to expose data about
+all changes that the control plane is making to external systems, with as much
+detail as possible so the platform team can understand what is changing, for
+what reason it is changing, and the outcome of the change. With this
+information, Crossplane users will be able to build deeper trust and confidence
+in the operation of their control planes, as well as have forensic data
+available to them in the event that an unexpected change occurs.
+
+## Proposal
+
+### Change Log Entry Format
+
+The change log will be a series of individual entries that capture state about
+the resources involved in every operation a provider performs on its
+external system. Later in this document we will explore HOW the logs are
+generated and stored, but for now we will focus on WHAT will be collected.
+
+As a general principal, we will be collecting raw observation data that has not
+been processed in any significant way. A system of the users choosing that
+collects the logs entries can perform opinionated calculations, analysis, or
+alerting that is specific to their operating environment.
+
+Each entry in the change log will contain the following data:
+
+| Field Description | Example |
+|-------------------| ------- |
+| Timestamp | ISO 8601 format, e.g. `2023-04-01T12:34:56Z` |
+| Provider name | `xpkg.upbound.io/upbound/provider-aws-ec2:v1.8.0` |
+| Managed Resource identifying data (GVK/name) | `{apiVersion: ec2.aws.upbound.io/v1beta2, kind: Instance, name: dev-instance-bt66d}` |
+| Operation Type | `create\|update\|delete` |
+| Before operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider` |
+| After operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider`  |
+| Result of operation | `success` or error object |
+| (optional) Additional information that Providers can set as they choose | `{JSON object of arbitrary data}` |
```

Comment:
I remember there was a separate discussion with @luxas on whether we can use the change logs feature to record data about all the reconciliations (for example, reconciliation timestamps), and not just about those reconciliations which attempt to correct drifts. Then the after observation would become optional (only recorded for those reconciliations that update the external resource). crossplane-runtime may not need to be opinionated on this, we would extend crossplane-runtime, and if needed, the package manager and the like, so that the side car (the gRPC server) and the client are properly setup, and the provider could use this infrastructure to implement some related features. The proposed arbitrary data field opens up the possibility for similar cases but the crossplane-runtime had better be un-opinionated about them, if we would like to allow other use cases.

Re-observing for the after-update external resource state would be the issue in this case, if we take the control from the managed reconciler. I think we would not like to have the external client to make a second observation on its own. One idea could be to allow the external client to be able to signal to the managed reconciler the need for an immediate reconciliation so that the managed reconciler requeues an immediate rate limited reconciliation request without waiting for the next poll period.

How do you folks feel about allowing the providers to record more reconciliation data, not limited to updates?

Reply:
> How do you folks feel about allowing the providers to record more reconciliation data, not limited to updates?

I don't feel like I understand well enough to answer the question. What are some examples of reconciliation data other than timestamps?